File: app/src/main/java/com/anokmik/tripassistant/trip/details/TripDetailsContract.java
package com.anokmik.tripassistant.trip.details;

import android.support.annotation.LayoutRes;

import com.anokmik.persistence.model.Trip;
import com.anokmik.persistence.model.TripEvent;
import com.anokmik.tripassistant.databinding.adapter.ViewHolderPresenter;

import java.util.List;

public interface TripDetailsContract {

    interface Presenter {

        List<TripEvent> getTripEvents();

        ViewHolderPresenter<TripEvent> getViewHolderPresenter();

        void showStartDatePicker();

        void showFinishDatePicker();

        void setStartDate(long startDate);

        void setFinishDate(long finishDate);

        boolean validFields();

        void save();

        void cancel();

        void edit();

        void delete();

    }

    interface TripEventListener {

        void addEvent();

        void deleteEvent(int position);

    }

    interface View {

        @LayoutRes
        int getRowItemLayoutId();

        int getItemBindingId();

        int getAdapterPositionProviderBindingId();

        int getItemListenerBindingId();

        int getItemIsEditingBindingId();

        void showStartDatePickerDialog(long startDate);

        void showFinishDatePickerDialog(long finishDate);

        void showDatesInvalidError();

        void addTripEvent();

        void viewTripEvent(long tripEventId);

        void enableSaveControls();

        void enableEditControls();

        void back();

    }

}

File: app/src/main/kotlin/com/anokmik/tripassistant/trip/details/TripDetailsContract.kt
package com.anokmik.tripassistant.trip.details

import com.anokmik.persistence.model.TripEvent
import com.anokmik.tripassistant.databinding.adapter.OnItemClickListener
import com.anokmik.tripassistant.databinding.adapter.ViewHolderPresenter
import com.anokmik.tripassistant.dialog.DateHandler
import com.anokmik.tripassistant.util.contract.presenter.PresenterEditable
import com.anokmik.tripassistant.util.contract.presenter.PresenterStartFinishDates
import com.anokmik.tripassistant.util.contract.presenter.PresenterValidator
import com.anokmik.tripassistant.util.contract.view.ViewBack
import com.anokmik.tripassistant.util.contract.view.ViewControls
import com.anokmik.tripassistant.util.contract.view.ViewListItem
import com.anokmik.tripassistant.util.contract.view.ViewStartFinishDates

interface TripDetailsContract {

    interface Presenter : TripEventListener, PresenterStartFinishDates, PresenterValidator, PresenterEditable, OnItemClickListener<TripEvent> {

        val tripEvents: MutableList<TripEvent>

        val viewHolderPresenter: ViewHolderPresenter<TripEvent>

    }

    interface View : ViewListItem, ViewStartFinishDates, ViewControls, ViewBack, DateHandler {

        fun addTripEvent()

        fun viewTripEvent(tripEventId: Long)

    }

}